libraries/rush-lib/src/cli/SelectionParameterSet.ts

```diff
@@ -401,4 +420,154 @@ export class SelectionParameterSet {
+     *
+     * When --to A is passed, the filter arguments are:
+     * 1. A is a split workspace project. C, G are covered by A.
+     *   splitWorkspace: --filter A...
+     * 2. B, F are workspace projects. D, E are covered by B.
+     *   workspace: --filter B... --filter F...
+     *
+     * NOTE: Workspace project can **NOT** depend on a split workspace project.
+     */
+    for (const project of projects) {
+      if (visited.has(project)) {
+        continue;
+      }
+      if (project.splitWorkspace) {
+        splitWorkspacePnpmFilterArguments.push(
+          '--filter',
+          `${project.packageName}${excludeSelf ? '^' : ''}...`
+        );
+        if (!excludeSelf) {
+          patched.add(project);
+        }
+        const queue: RushConfigurationProject[] = [project];
+        while (queue.length > 0) {
+          const project: RushConfigurationProject = queue.shift() as RushConfigurationProject;
+          const { dependencyProjects } = project;
+          for (const dependencyProject of dependencyProjects) {
+            if (dependencyProject.splitWorkspace) {
+              if (!visited.has(dependencyProject)) {
+                queue.push(dependencyProject);
+              }
+            } else {
+              if (!patched.has(dependencyProject)) {
+                pnpmFilterArguments.push('--filter', `${dependencyProject.packageName}...`);
+                patched.add(dependencyProject);
+              }
+            }
+            visited.add(dependencyProject);
+          }
+        }
+      } else {
+        pnpmFilterArguments.push('--filter', `${project.packageName}${excludeSelf ? '^' : ''}...`);
+        if (!excludeSelf) {
+          patched.add(project);
+        }
+      }
+      visited.add(project);
+    }
+  }
+
+  private _pushPnpmFilterArgumentsByFromProjects({
+    projects,
+    pnpmFilterArguments,
+    splitWorkspacePnpmFilterArguments,
+    excludeSelf
+  }: {
+    projects: Set<RushConfigurationProject>;
+    pnpmFilterArguments: string[];
+    splitWorkspacePnpmFilterArguments: string[];
+    excludeSelf: boolean;
+  }): void {
+    if (projects.size === 0) {
+      return;
+    }
+    const visited: Set<RushConfigurationProject> = new Set<RushConfigurationProject>();
+    const patched: Set<RushConfigurationProject> = new Set<RushConfigurationProject>();
+    /**
+     * Assuming:
+     *
+     *       A - splitWorkspace
+     *      / \
+     *     B   C - splitWorkspace
+     *    / \ / \
+     *   D  E F  G - splitWorkspace
+     *
+     * When --from G is passed, G is a split workspace project, so all its
+     * dependents are covered from G
+     *    splitWorkspace: --filter G...
+     *
+     * When --from D is passed,
+     * 1. D is workspace project. B is covered by D.
+     *    workspace: --filter ...D
+     * 2. A is a split workspace project.
+     *    splitWorkspace: --filter ...A
+     *
+     * NOTE: Workspace project can **NOT** depend on a split workspace project.
```

Comment:
One concern is that expanding all selection into `--filter` make the args string pretty long. Is there any chance that we run into shell args length limitation?

Like, Windows platform limits 8191 chars as max length. 
https://docs.microsoft.com/en-US/troubleshoot/windows-client/shell-experience/command-line-string-limitation

Reply:
Your latest iteration is definitely simpler. Regarding the 8191 limit, Zoltan confirmed that PNPM does not provide a way to read CLI parameters from a text file input. It does not directly affect us because our monorepo does not currently build on Windows. However if we want to support that, one option would be for Rush to make a small wrapper that assigns `process.argv` manually and then loads PNPM by doing `require('./node_modules/pnpm/bin/pnpm.cjs')`.  (Another option is to contribute an `--args-from-file` feature to PNPM.  It's interesting that Node.exe itself does not have this feature.)